### PR TITLE
Discrete integration slider in the instrument viewer

### PR DESCRIPTION
**Pull request by MathieuTillet**
**Description of work.**
This PR modifies the integration slider of the instrument viewer so that if the bin axis of the workspace is discrete, the slider is not continuous but works in discrete intervals based on the indices of the workspace's bins. 

**To test:**
-  Load `./ExternalData/Testing/Data/UnitTest/ILL/D11B/kinetic.nxs` and open the instrument viewer.
- Check the slider goes from 0 to 2.
- Try moving the slider, reducing and increasing its size, both with the slider and the spin boxes, and both with mouse and keyboard,  and check the view is updated accordingly. 

WIP for #31943. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/src/Algorithm.cpp [hunk 1]
```diff
@@ -737,16 +737,18 @@ bool Algorithm::executeInternal() {
...
       (Mantid::Types::Core::DateAndTime::ONE_SECOND * DELAY_BEFORE_GC);
   if (algIsExecuted) {
     setResultState(ResultState::Success);
   }
+
   // Only gets to here if algorithm ended normally
   notificationCenter().postNotification(new FinishedNotification(this, isExecuted()));
 
+  // Unlock the workspaces once the notification has been sent, to prevent too early deletion
+  this->unlockWorkspaces();
```
Comment (gvardany) [inline Framework/API/src/Algorithm.cpp:750]: 👍

**Review by gvardany — COMMENTED** [review]
This works mostly fine. If the slider is discrete, the default slider button should be of width 1 positioned at index 0, that is, the first frame, instead of the integral. The other major issue is that the arrow keys are ignored on macOS, we will have to investigate why. A minor issue is that when editing the start or end via spin-boxes, one has to click out to get the effect; just changing the spin-box value doesn't trigger the view update.

**Comment by gvardany** [comment]
It seems the key press issue was trivial.

**Comment by MathieuTillet** [comment]
The spin box not triggering the whole integration process for every single value it gets is intentional, since it would be excessively slow.

**Comment by gvardany** [comment]
Got a conflict now, I think this file was moved.

File: qt/widgets/instrumentview/src/InstrumentWidget.cpp [hunk 2]
```diff
@@ -612,18 +612,36 @@ void InstrumentWidget::replaceWorkspace(const std::string &newWs, const std::str
...
+ * @param init : boolean set to true if the integration widget is still being initialized
  */
 void InstrumentWidget::updateIntegrationWidget(bool init) {
-  m_xIntegration->setTotalRange(m_instrumentActor->minWkspBinValue(), m_instrumentActor->maxWkspBinValue());
+  // discrete integration range is only used if all the bins are common and integers and not an event workspace, as a
+  // convention
+  auto ws = m_instrumentActor->getWorkspace();
+
+  // differentiating behaviour by subclass is not really recommended, but we have to chek if there are events.
+  bool isNotEventWs = dynamic_cast<const Mantid::DataObjects::EventWorkspace *>(ws.get()) == nullptr;
```
Comment (gvardany) [inline qt/widgets/instrumentview/src/InstrumentWidget.cpp:624]: Can you use the `id()` method to avoid the cast?

**Comment by gvardany** [comment]
Discrete slider works well. I just noticed that for the continuous slider, I'm not able to move/resize the scroll button with arrow key press on mac (with discrete, however, it works). Can you check this on linux? I think this is related to the text boxes, which are present only if continuous, and they seem to take the focus.

**Comment by gvardany** [comment]
> The spin box not triggering the whole integration process for every single value it gets is intentional, since it would be excessively slow.

This makes sense, but might not be very intuitive. Could you put a hint in tooltips that one needs to click out to take effect?

File: qt/widgets/instrumentview/src/InstrumentWidget.cpp [hunk 3]
```diff
@@ -612,18 +612,36 @@ void InstrumentWidget::replaceWorkspace(const std::string &newWs, const std::str
...
+ * Update the range of the integration widget, and show or hide it if needed
+ * @param init : boolean set to true if the integration widget is still being initialized
  */
 void InstrumentWidget::updateIntegrationWidget(bool init) {
-  m_xIntegration->setTotalRange(m_instrumentActor->minWkspBinValue(), m_instrumentActor->maxWkspBinValue());
+  // discrete integration range is only used if all the bins are common and integers and not an event workspace, as a
+  // convention
+  auto ws = m_instrumentActor->getWorkspace();
+
+  // differentiating behaviour by subclass is not really recommended, but we have to chek if there are events.
```
Comment (gvardany) [inline qt/widgets/instrumentview/src/InstrumentWidget.cpp:623]: This comment is not relevant anymore.

File: qt/widgets/instrumentview/src/InstrumentWidget.cpp [hunk 4]
```diff
@@ -23,6 +23,7 @@
 #include "MantidAPI/IPeaksWorkspace.h"
 #include "MantidAPI/MatrixWorkspace.h"
 #include "MantidAPI/Workspace.h"
+#include "MantidDataObjects/EventWorkspace.h"
```
Comment (gvardany) [inline qt/widgets/instrumentview/src/InstrumentWidget.cpp:26]: Now that you don't need the complete type, you can change this to the forward declaration: `EventWorkspace_fwd.h`.

Comment (gvardany) [reply qt/widgets/instrumentview/src/InstrumentWidget.cpp:26]: Actually, you can remove it altogether.

**Comment by gvardany** [comment]
This is good now, just merge master once PR #32589 is in.

**Review by gvardany — APPROVED** [review]

**Pull request merged**